dist/platforms/mac/steps/activate.sh

```diff
@@ -6,20 +6,41 @@ pushd "$ACTIVATE_LICENSE_PATH"
 
 echo "Requesting activation"
 
-# Activate license
-/Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/MacOS/Unity \
-  -logFile - \
-  -batchmode \
-  -nographics \
-  -quit \
-  -serial "$UNITY_SERIAL" \
-  -username "$UNITY_EMAIL" \
-  -password "$UNITY_PASSWORD" \
-  -projectPath "$ACTIVATE_LICENSE_PATH"
+if [[ -n "$UNITY_LICENSING_SERVER" ]]; then
+  #
+  # Custom Unity License Server
+  #
+  echo "Adding licensing server config"
+  mkdir -p "/Library/Application Support/Unity/config/"
+  cp \
+    "$ACTION_FOLDER/unity-config/services-config.json" \
+    "/Library/Application Support/Unity/config/services-config.json"
+  /Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/Frameworks/UnityLicensingClient.app/Contents/MacOS/Unity.Licensing.Client \
+    --acquire-floating > license.txt
+  PARSEDFILE=$(grep -oE '\".*?\"' < license.txt | tr -d '"')
+  grep -oE '\".*?\"' < license.txt
+  export FLOATING_LICENSE
+  FLOATING_LICENSE=$(sed -n 2p <<< "$PARSEDFILE")
+  FLOATING_LICENSE_TIMEOUT=$(sed -n 4p <<< "$PARSEDFILE")
 
-# Store the exit code from the verify command
-UNITY_EXIT_CODE=$?
+  echo "Acquired floating license: \"$FLOATING_LICENSE\" with timeout $FLOATING_LICENSE_TIMEOUT"
+  # Store the exit code from the verify command
+  UNITY_EXIT_CODE=$?
```

Comment:
This is not exactly correct because `$?` here will store the exit code from the above `echo` command. Maybe it should be placed bellow `/Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/Frameworks/UnityLicensingClient.app/Contents/MacOS/Unity.Licensing.Client` command, but it's also different from `/Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/MacOS/Unity` so I think there should be a separate control flow for validating the success of the `UNITY_LICENSING_SERVER` activation mechanism.

Reply:
Here's a proposition (untested):

```bash
#!/usr/bin/env bash

# Run in ACTIVATE_LICENSE_PATH directory
echo "Changing to \"$ACTIVATE_LICENSE_PATH\" directory."
pushd "$ACTIVATE_LICENSE_PATH"

echo "Requesting activation"

if [[ -n "$UNITY_LICENSING_SERVER" ]]; then
  #
  # Custom Unity License Server
  #
  echo "Adding licensing server config"
  mkdir -p "/Library/Application Support/Unity/config/"
  cp \
    "$ACTION_FOLDER/unity-config/services-config.json" \
    "/Library/Application Support/Unity/config/services-config.json"
  /Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/Frameworks/UnityLicensingClient.app/Contents/MacOS/Unity.Licensing.Client \
    --acquire-floating > license.txt
  
  # Store the exit code from the acquire-floating command
  UNITY_LICENSING_EXIT_CODE=$?

  PARSEDFILE=$(grep -oE '\".*?\"' < license.txt | tr -d '"')
  grep -oE '\".*?\"' < license.txt
  export FLOATING_LICENSE
  FLOATING_LICENSE=$(sed -n 2p <<< "$PARSEDFILE")
  FLOATING_LICENSE_TIMEOUT=$(sed -n 4p <<< "$PARSEDFILE")

  echo "Acquired floating license: \"$FLOATING_LICENSE\" with timeout $FLOATING_LICENSE_TIMEOUT"
else
  # Activate license
  /Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/MacOS/Unity \
    -logFile - \
    -batchmode \
    -nographics \
    -quit \
    -serial "$UNITY_SERIAL" \
    -username "$UNITY_EMAIL" \
    -password "$UNITY_PASSWORD" \
    -projectPath "$ACTIVATE_LICENSE_PATH"

  # Store the exit code from the verify command
  UNITY_LICENSE_EXIT_CODE=$?
fi
#
# Display information about the result
#
if [[ -n "$UNITY_LICENSING_SERVER" ]]; then
  if [ $UNITY_LICENSING_EXIT_CODE -eq 0 ]; then
    # Activation was a success
    echo "Activation complete."
  else
    # Activation failed so exit with the code from the license verification step
    echo "Unclassified error occurred while trying to activate license with UNITY_LICENSING_SERVER."
    echo "Exit code was: $UNITY_LICENSING_EXIT_CODE"
    exit $UNITY_LICENSING_EXIT_CODE
  fi
else
  if [ $UNITY_LICENSE_EXIT_CODE -eq 0 ]; then
    # Activation was a success
    echo "Activation complete."
  else
    # Activation failed so exit with the code from the license verification step
    echo "Unclassified error occurred while trying to activate license."
    echo "Exit code was: $UNITY_LICENSE_EXIT_CODE"
    exit $UNITY_LICENSE_EXIT_CODE
  fi
fi

# Return to previous working directory
popd
```

Reply:
You are right, I can see it is wrong in the ubuntu version also, I will add alternative way of verifying it

Reply:
I went ahead and asked ChatGPT to improve the script and it came out with this (untested):

```bash
#!/usr/bin/env bash

check_exit_code() {
  local exit_code=$1
  local error_message=$2

  if [ $exit_code -eq 0 ]; then
    echo "Activation complete."
  else
    echo "Unclassified error occurred while trying to activate license."
    echo "$error_message"
    echo "Exit code was: $exit_code"
    exit $exit_code
  fi
}

echo "Changing to \"$ACTIVATE_LICENSE_PATH\" directory."
if ! cd "$ACTIVATE_LICENSE_PATH"; then
  echo "Failed to change directory to \"$ACTIVATE_LICENSE_PATH\"."
  exit 1
fi

echo "Requesting activation"

if [[ -n "$UNITY_LICENSING_SERVER" ]]; then
  echo "Adding licensing server config"
  mkdir -p "/Library/Application Support/Unity/config/"
  cp "$ACTION_FOLDER/unity-config/services-config.json" "/Library/Application Support/Unity/config/services-config.json"

  /Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/Frameworks/UnityLicensingClient.app/Contents/MacOS/Unity.Licensing.Client --acquire-floating > license.txt
  unity_exit_code=$?
  
  parsed_file=$(grep -oE '\".*?\"' < license.txt | tr -d '"')
  grep -oE '\".*?\"' < license.txt
  export FLOATING_LICENSE
  FLOATING_LICENSE=$(sed -n 2p <<< "$parsed_file")
  FLOATING_LICENSE_TIMEOUT=$(sed -n 4p <<< "$parsed_file")

  echo "Acquired floating license: \"$FLOATING_LICENSE\" with timeout $FLOATING_LICENSE_TIMEOUT"
else
  /Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/MacOS/Unity -logFile - -batchmode -nographics -quit -serial "$UNITY_SERIAL" -username "$UNITY_EMAIL" -password "$UNITY_PASSWORD" -projectPath "$ACTIVATE_LICENSE_PATH"
  unity_exit_code=$?
fi

if [[ -n "$UNITY_LICENSING_SERVER" ]]; then
  check_exit_code $unity_exit_code "Error occurred while trying to activate license with UNITY_LICENSING_SERVER."
else
  check_exit_code $unity_exit_code "Error occurred while trying to activate license."
fi

# Return to previous working directory
popd
```

Might be worth trying

Reply:
Then I asked for a few more functions as this usually makes it more readable :)

```bash
#!/usr/bin/env bash

check_exit_code() {
  local exit_code=$1
  local error_message=$2

  if [ $exit_code -eq 0 ]; then
    echo "Activation complete."
  else
    echo "Unclassified error occurred while trying to activate license."
    echo "$error_message"
    echo "Exit code was: $exit_code"
    exit $exit_code
  fi
}

echo "Changing to \"$ACTIVATE_LICENSE_PATH\" directory."
if ! cd "$ACTIVATE_LICENSE_PATH"; then
  echo "Failed to change directory to \"$ACTIVATE_LICENSE_PATH\"."
  exit 1
fi

echo "Requesting activation"

if [[ -n "$UNITY_LICENSING_SERVER" ]]; then
  echo "Adding licensing server config"
  mkdir -p "/Library/Application Support/Unity/config/"
  cp "$ACTION_FOLDER/unity-config/services-config.json" \
     "/Library/Application Support/Unity/config/services-config.json"

  /Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/Frameworks/UnityLicensingClient.app/Contents/MacOS/Unity.Licensing.Client \
    --acquire-floating > license.txt
  unity_exit_code=$?
  
  parsed_file=$(grep -oE '\".*?\"' < license.txt | tr -d '"')
  grep -oE '\".*?\"' < license.txt
  export FLOATING_LICENSE
  FLOATING_LICENSE=$(sed -n 2p <<< "$parsed_file")
  FLOATING_LICENSE_TIMEOUT=$(sed -n 4p <<< "$parsed_file")

  echo "Acquired floating license: \"$FLOATING_LICENSE\" with timeout $FLOATING_LICENSE_TIMEOUT"
else
  /Applications/Unity/Hub/Editor/$UNITY_VERSION/Unity.app/Contents/MacOS/Unity \
    -logFile - \
    -batchmode \
    -nographics \
    -quit \
    -serial "$UNITY_SERIAL" \
    -username "$UNITY_EMAIL" \
    -password "$UNITY_PASSWORD" \
    -projectPath "$ACTIVATE_LICENSE_PATH"
  unity_exit_code=$?
fi

if [[ -n "$UNITY_LICENSING_SERVER" ]]; then
  check_exit_code $unity_exit_code "Error occurred while trying to activate license with UNITY_LICENSING_SERVER."
else
  check_exit_code $unity_exit_code "Error occurred while trying to activate license."
fi

# Return to previous working directory
popd
```

Reply:
Actually I'd argue that these functions don't really make the code more readable.

The following completely detaches the error checking and does it based on a condition that is unity licensing server, which adds cognitive complexity.

```
if [[ -n "$UNITY_LICENSING_SERVER" ]]; then
  check_exit_code $unity_exit_code "Error occurred while trying to activate license with UNITY_LICENSING_SERVER."
else
  check_exit_code $unity_exit_code "Error occurred while trying to activate license."
fi
```

I'd really prefer to keep the current paradigm for now if you don't mind, so that we can move these flows into the CLI (and run that CLI from runner / inside the container) more easily. These bash scripts should really go 😅

Reply:
yeah agreed, sorry for the noise :)

[The first proposition](https://github.com/game-ci/unity-builder/pull/573#discussion_r1330546993) that only adds `UNITY_LICENSING_EXIT_CODE=$?` is most likely better for now 👍

Reply:
Sounds good to me, feel free to change, if not I can do it tomorrow!